rfcs/0013-data-plane-fanout.md

```diff
@@ -0,0 +1,86 @@
+      |                |       | | |            +----------------------+ |                       |
+      |                |       | | |            |                        |                       |
+      |                |       | | +--------------------------------------------+                |
+      |                |       | |              |                        |      |                |
+      |                |       | +----------------------+                |      |                |
+      v                v       v                v       v                v      v                v
+    +-+----------------+-+   +-+----------------+-+   +-+----------------+-+  +-+----------------+-+
+    |                    |   |                    |   |                    |  |                    |
+    |   Neighborhood 1   |   |   Neighborhood 2   |   |   Neighborhood 3   |  |   Neighborhood 4   |
+    |                    |   |                    |   |                    |  |                    |
+    +--------------------+   +--------------------+   +--------------------+  +--------------------+
+              
+#### A Stake weighted selection mechanism
+To support this mechanism, there needs to be a agreed upon way of dividing the network amongst the nodes. To achieve this the `tvu_peers` are sorted by stake and stored in a list. This list can then be indexed in different ways to figure out neighborhood boundaries and retransmit peers.
+For example, the leader will simply select the first `DATA_PLANE_FANOUT` nodes as its layer 1 nodes. These will automatically be the highest stake holders allowing the heaviest votes to come back to the leader first. 
+The same logic determines which nodes each node in layer needs to retransmit its blobs to. This involves finding its neighbors and lower layer peers.    
+ 
+#### Broadcast stage
+Broadcast Stage uses a bank to figure out stakes and hands this off to ClusterInfo to figure out the top `DATA_PLANE_FANOUT` stake holders. 
+These top stake holders will be considered Layer 1. For the leader this is pretty straightforward and can be achieved with a `truncate` call on a sorted list of peers. 
+
+#### Replicate stage
+The biggest challenge in updating to this mechanism is to update replicate stage and make it "layer aware"; i.e using the bank each node can figure out which layer it belongs in and which lower layer peers nodes to send blobs to with minimal overlap across neighborhood boundaries. 
+Overlaps will be minimized based on `((node.layer_index) % (layer_neighborhood_size) * cur_neighborhood_index)` where `cur_neighborhood_index` is the loop index in `num_neighborhoods` so that a node only forwards blobs to a single node in a lower layer neighborhood. 
+
+Each node can receive blobs froms its peer in the layer above as well as its neighbors. As long as the failure rate is less than the number of erasure codes, blobs can be repaired without the network failing. 
```

Comment:
Each neighbor has a different peer from the layer above?

Reply:
they see the layer above as any number of leaders, I think

Reply:
@sagar-solana Can you elaborate that part?  Each neighbor with different leaders + erasure codes it what keeps any set of leaders from blocking the neighborhood.

Reply:
Sure, so a node in Layer X will receive blobs from "above" (layer X-1), it doesn't care what is above, it could be the leader directly or another node in a higher layer neighborhood. When it receives that blob, it assumes(if it's a leader blob, based on signature) that it needs to forward it to its peers. All nodes that are in its neighborhood and a single node from each neighborhood in a lower layer (Layer X+1) are selected as retransmit peers. I'm assuming this part is relatively clear.

So, for any node N in a layer X, it receives blobs from a node in each neighborhood from a layer above, and retransmits it across the neighborhood and below to a lower layer peer. For any given neighborhood in a layer, each node in that neighborhood could be receiving _different_ blobs from above and because of the retransmit will end up sharing them with its neighbors (similar to how a leader is round-robin'ing blobs to layer 1, requiring layer 1 nodes to retransmit amongst themselves) leading to a completed window. To cripple a neighborhood, enough nodes (erasure codes +1 _per neighborhood_) from the layer above need to fail. Since multiple neighborhoods can exist in the upper layer and a node will receive blobs from a node in each of those neighborhoods, you'd have to have a whole bunch of failures on the upper layers to end up with incomplete windows.

I hope that makes a little more sense? I almost confused myself explaining it.

Reply:
I sort of have that already ;) The code that figures out neighbors and peers doesn't do any IO it's just a couple of math functions that figure out how to slice an array of nodes to figure out layer boundaries and peers. I've got functions to test 100% 'network' coverage, but I haven't tested 100% coverage if I randomly disable indices or skip over them(in theory, like explained above, I rely on retransmits to fill the gaps). I'm not sure how to make sure a blob travels the whole network properly without doing any actual IO. 

What do you have in mind? Maybe I can try that.

Reply:
crds_gossip module simulates a large network without any IO. 

The property we care about is that of the percentage of erasure codes is greater than failed nodes that the vast majority of the network will be able to reconstruct the data without using the repair mechanism.